# Feature Engineering Team - EDA-Team-1

### Contributers:
#### 1. [Rahul Mandviya (Lead)](https://github.com/Rahulm0106)
#### 2. [Rohan Jamdade](https://github.com/rjrohan13)
#### 3. [Ekaterine](https://github.com/Ekaterine-git)
#### 4. [Anastasiia Marchenko](https://github.com/dots13)
#### 5. [Tamil Arsu](https://github.com/Tamilarasu16)
#### 6. [Sabelo Mbatha](https://github.com/Sabunity)
#### 7. [Juhee Jain](https://github.com/JJAIN02)
#### 8. [Sudarshan](https://github.com/sudharsanBabu)
#### 9. [Gladys Kiruba]()

In [1]:


# importing libraries
import pandas as pd
import numpy as np
from scipy.stats import f_oneway
from scipy.stats import chi2_contingency
import scipy.stats as stats

In [2]:
df = pd.read_csv('Final-data.csv')

In [3]:
df

Unnamed: 0 school sex  age address famsize Pstatus  Medu  Fedu  \
0              0     GP   F   18       U     GT3       A     4     4   
1              1     GP   F   17       U     GT3       T     1     1   
2              2     GP   F   15       U     LE3       T     1     1   
3              3     GP   F   15       U     GT3       T     4     2   
4              4     GP   F   16       U     GT3       T     3     3   
...          ...    ...  ..  ...     ...     ...     ...   ...   ...   
1039        1039     MS   F   19       R     GT3       T     2     3   
1040        1040     MS   F   18       U     LE3       T     3     1   
1041        1041     MS   F   18       U     GT3       T     1     1   
1042        1042     MS   M   17       U     LE3       T     3     1   
1043        1043     MS   M   18       R     LE3       T     3     2   

          Mjob  ... freetime goout Dalc  Walc  health  absences  G1  G2  G3  \
0      at_home  ...        3     4    1     1       3         6   5   6   6   
1      at_home  ...        3     3    1     1       3         4   5   5   6   
2      at_home  ...        3     2    2     3       3        10   7   8  10   
3       health  ...        2     2    1     1       5         2  15  14  15   
4        other  ...        3     2    1     2       5         4   6  10  10   
...        ...  ...      ...   ...  ...   ...     ...       ...  ..  ..  ..   
1039  services  ...        4     2    1     2       5         4  10  11  10   
1040   teacher  ...        3     4    1     1       1         4  15  15  16   
1041     other  ...        1     1    1     1       5         6  11  12   9   
1042  services  ...        4     5    3     4       2         6  10  10  10   
1043  services  ...        4     1    3     4       5         4  10  11  11   

     Subject  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
1039       0  
1040       0  
1041       0  
1042       0  
1043       0  

[1044 rows x 35 columns]

## Derived Feature - Final_Score

#### 30% weightage is given to period 1 and period 2 exam and 40% is given to final exam.
#### These weights have been based on correlations between G1, G2, G3, Final_Score.

In [4]:
weight_g1 = 0.3
weight_g2 = 0.3
weight_g3 = 0.4

In [5]:
df["Final_Score"] = (df['G1'] * weight_g1) + (df['G2'] * weight_g2) + (df['G3'] * weight_g3)
df["Final_Score"] = df["Final_Score"].round()
df

Unnamed: 0 school sex  age address famsize Pstatus  Medu  Fedu  \
0              0     GP   F   18       U     GT3       A     4     4   
1              1     GP   F   17       U     GT3       T     1     1   
2              2     GP   F   15       U     LE3       T     1     1   
3              3     GP   F   15       U     GT3       T     4     2   
4              4     GP   F   16       U     GT3       T     3     3   
...          ...    ...  ..  ...     ...     ...     ...   ...   ...   
1039        1039     MS   F   19       R     GT3       T     2     3   
1040        1040     MS   F   18       U     LE3       T     3     1   
1041        1041     MS   F   18       U     GT3       T     1     1   
1042        1042     MS   M   17       U     LE3       T     3     1   
1043        1043     MS   M   18       R     LE3       T     3     2   

          Mjob  ... goout Dalc Walc  health  absences  G1  G2  G3 Subject  \
0      at_home  ...     4    1    1       3         6   5   6   6       1   
1      at_home  ...     3    1    1       3         4   5   5   6       1   
2      at_home  ...     2    2    3       3        10   7   8  10       1   
3       health  ...     2    1    1       5         2  15  14  15       1   
4        other  ...     2    1    2       5         4   6  10  10       1   
...        ...  ...   ...  ...  ...     ...       ...  ..  ..  ..     ...   
1039  services  ...     2    1    2       5         4  10  11  10       0   
1040   teacher  ...     4    1    1       1         4  15  15  16       0   
1041     other  ...     1    1    1       5         6  11  12   9       0   
1042  services  ...     5    3    4       2         6  10  10  10       0   
1043  services  ...     1    3    4       5         4  10  11  11       0   

     Final_Score  
0            6.0  
1            5.0  
2            8.0  
3           15.0  
4            9.0  
...          ...  
1039        10.0  
1040        15.0  
1041        10.0  
1042        10.0  
1043        11.0  

[1044 rows x 36 columns]

In [6]:
df = df.drop(['Unnamed: 0','G1','G2','G3'], axis = 1)

In [7]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0        GP   F   18       U     GT3       A     4     4   at_home   teacher   
1        GP   F   17       U     GT3       T     1     1   at_home     other   
2        GP   F   15       U     LE3       T     1     1   at_home     other   
3        GP   F   15       U     GT3       T     4     2    health  services   
4        GP   F   16       U     GT3       T     3     3     other     other   
...     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
1039     MS   F   19       R     GT3       T     2     3  services     other   
1040     MS   F   18       U     LE3       T     3     1   teacher  services   
1041     MS   F   18       U     GT3       T     1     1     other     other   
1042     MS   M   17       U     LE3       T     3     1  services  services   
1043     MS   M   18       R     LE3       T     3     2  services     other   

      ... romantic famrel  freetime  goout  Dalc Walc health absences Subject  \
0     ...       no      4         3      4     1    1      3        6       1   
1     ...       no      5         3      3     1    1      3        4       1   
2     ...       no      4         3      2     2    3      3       10       1   
3     ...      yes      3         2      2     1    1      5        2       1   
4     ...       no      4         3      2     1    2      5        4       1   
...   ...      ...    ...       ...    ...   ...  ...    ...      ...     ...   
1039  ...       no      5         4      2     1    2      5        4       0   
1040  ...       no      4         3      4     1    1      1        4       0   
1041  ...       no      1         1      1     1    1      5        6       0   
1042  ...       no      2         4      5     3    4      2        6       0   
1043  ...       no      4         4      1     3    4      5        4       0   

     Final_Score  
0            6.0  
1            5.0  
2            8.0  
3           15.0  
4            9.0  
...          ...  
1039        10.0  
1040        15.0  
1041        10.0  
1042        10.0  
1043        11.0  

[1044 rows x 32 columns]

## Derived Feature - SES_Status

In [8]:
# Function to map socioeconomic status based on features which could affect ses
def get_ses_status_score(row):
    # Calculate the SES score based on relevant features
    ses_score = 0

    # Add points based on parents' education
    if row['Medu'] in ['0', '1']:
        ses_score += 1
    elif row['Medu'] == '2':
        ses_score += 2
    elif row['Medu'] == '3':
        ses_score += 3
    elif row['Medu'] == '4':
        ses_score += 4

    if row['Fedu'] in ['0', '1']:
        ses_score += 1
    elif row['Fedu'] == '2':
        ses_score += 2
    elif row['Fedu'] == '3':
        ses_score += 3
    elif row['Fedu'] == '4':
        ses_score += 4

    # Add points based on parents' occupation (Mother_job and Father_job)
    if row['Mjob'] in ['at_home', 'other']:
        ses_score += 1
    elif row['Mjob'] in ['teacher', 'health', 'services']:
        ses_score += 2

    if row['Fjob'] in ['at_home', 'other']:
        ses_score += 1
    elif row['Fjob'] in ['teacher', 'health', 'services']:
        ses_score += 2

    # Add points based on family size (famsize)
    if row['famsize'] == 'LE3':
        ses_score += 1
        
    # Add points based on address (Urban or Rural)
    if row['address'] == 'U':
        ses_score += 1

    # Add points based on Parents_status (Together or Apart)
    if row['Pstatus'] == 'T':
        ses_score += 1
        
    # Add points based on guardian
    if row['guardian'] in ['mother', 'father']:
        ses_score += 1
    elif row['guardian'] == 'other':
        ses_score += 2

    # Add points based on school_support, fam_support, paid_extra_classes, higher_edu, internet, and nursery
    if row['schoolsup'] == 'yes':
        ses_score += 1

    if row['famsup'] == 'yes':
        ses_score += 1

    if row['paid'] == 'yes':
        ses_score += 1

    if row['higher'] == 'yes':
        ses_score += 1

    if row['internet'] == 'yes':
        ses_score += 1

    if row['nursery'] == 'yes':
        ses_score += 1

    return ses_score

In [9]:
df['ses_score'] = df.apply(get_ses_status_score, axis=1)
df.head(20)

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0      GP   F   18       U     GT3       A     4     4   at_home   teacher   
1      GP   F   17       U     GT3       T     1     1   at_home     other   
2      GP   F   15       U     LE3       T     1     1   at_home     other   
3      GP   F   15       U     GT3       T     4     2    health  services   
4      GP   F   16       U     GT3       T     3     3     other     other   
5      GP   M   16       U     LE3       T     4     3  services     other   
6      GP   M   16       U     LE3       T     2     2     other     other   
7      GP   F   17       U     GT3       A     4     4     other   teacher   
8      GP   M   15       U     LE3       A     3     2  services     other   
9      GP   M   15       U     GT3       T     3     4     other     other   
10     GP   F   15       U     GT3       T     4     4   teacher    health   
11     GP   F   15       U     GT3       T     2     1  services     other   
12     GP   M   15       U     LE3       T     4     4    health  services   
13     GP   M   15       U     GT3       T     4     3   teacher     other   
14     GP   M   15       U     GT3       A     2     2     other     other   
15     GP   F   16       U     GT3       T     4     4    health     other   
16     GP   F   16       U     GT3       T     4     4  services  services   
17     GP   F   16       U     GT3       T     3     3     other     other   
18     GP   M   17       U     GT3       T     3     2  services  services   
19     GP   M   16       U     LE3       T     4     3    health     other   

    ... famrel freetime  goout  Dalc  Walc health absences Subject  \
0   ...      4        3      4     1     1      3        6       1   
1   ...      5        3      3     1     1      3        4       1   
2   ...      4        3      2     2     3      3       10       1   
3   ...      3        2      2     1     1      5        2       1   
4   ...      4        3      2     1     2      5        4       1   
5   ...      5        4      2     1     2      5       10       1   
6   ...      4        4      4     1     1      3        0       1   
7   ...      4        1      4     1     1      1        6       1   
8   ...      4        2      2     1     1      1        0       1   
9   ...      5        5      1     1     1      5        0       1   
10  ...      3        3      3     1     2      2        0       1   
11  ...      5        2      2     1     1      4        4       1   
12  ...      4        3      3     1     3      5        2       1   
13  ...      5        4      3     1     2      3        2       1   
14  ...      4        5      2     1     1      3        0       1   
15  ...      4        4      4     1     2      2        4       1   
16  ...      3        2      3     1     2      2        6       1   
17  ...      5        3      2     1     1      4        4       1   
18  ...      5        5      5     2     4      5       16       1   
19  ...      3        1      3     1     3      5        4       1   

   Final_Score ses_score  
0          6.0         8  
1          5.0         8  
2          8.0        11  
3         15.0        12  
4          9.0         9  
5         15.0        12  
6         12.0         9  
7          6.0         9  
8         18.0        11  
9         15.0        10  
10         9.0        12  
11        11.0        10  
12        14.0        13  
13        10.0        11  
14        15.0         9  
15        14.0        10  
16        14.0        12  
17         9.0         9  
18         5.0        11  
19         9.0        11  

[20 rows x 33 columns]

In [10]:
# Calculate quartiles to derive threshold values
q1 = np.percentile(df['ses_score'], 25)
q2 = np.percentile(df['ses_score'], 50)  # Median
q3 = np.percentile(df['ses_score'], 75)

# Print the calculated quartiles
print("1st Quartile (25th percentile):", q1)
print("2nd Quartile (Median, 50th percentile):", q2)
print("3rd Quartile (75th percentile):", q3)

1st Quartile (25th percentile): 8.0
2nd Quartile (Median, 50th percentile): 9.0
3rd Quartile (75th percentile): 10.0


In [11]:
ses_array = []
ses_array = df['ses_score']
ses_array = set(ses_array)
ses_array

{5, 6, 7, 8, 9, 10, 11, 12, 13}

In [12]:
def get_ses_status_category(row):
    # Categorize SES score into 'Low', 'Medium', and 'High'
    if row['ses_score'] < 8:
        return 'Low'
    elif row['ses_score'] <= 10:
        return 'Medium'
    else:
        return 'High'

In [13]:
df['ses_status'] = df.apply(get_ses_status_category, axis=1)

In [14]:
df.head(10)

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4   at_home   teacher   
1     GP   F   17       U     GT3       T     1     1   at_home     other   
2     GP   F   15       U     LE3       T     1     1   at_home     other   
3     GP   F   15       U     GT3       T     4     2    health  services   
4     GP   F   16       U     GT3       T     3     3     other     other   
5     GP   M   16       U     LE3       T     4     3  services     other   
6     GP   M   16       U     LE3       T     2     2     other     other   
7     GP   F   17       U     GT3       A     4     4     other   teacher   
8     GP   M   15       U     LE3       A     3     2  services     other   
9     GP   M   15       U     GT3       T     3     4     other     other   

   ... freetime goout  Dalc  Walc  health absences Subject Final_Score  \
0  ...        3     4     1     1       3        6       1         6.0   
1  ...        3     3     1     1       3        4       1         5.0   
2  ...        3     2     2     3       3       10       1         8.0   
3  ...        2     2     1     1       5        2       1        15.0   
4  ...        3     2     1     2       5        4       1         9.0   
5  ...        4     2     1     2       5       10       1        15.0   
6  ...        4     4     1     1       3        0       1        12.0   
7  ...        1     4     1     1       1        6       1         6.0   
8  ...        2     2     1     1       1        0       1        18.0   
9  ...        5     1     1     1       5        0       1        15.0   

  ses_score ses_status  
0         8     Medium  
1         8     Medium  
2        11       High  
3        12       High  
4         9     Medium  
5        12       High  
6         9     Medium  
7         9     Medium  
8        11       High  
9        10     Medium  

[10 rows x 34 columns]

In [15]:
#dropping the features used in calculating ses_status
df = df.drop(['address', 'famsize' , 'Pstatus','Medu','Fedu','Mjob','Fjob',
            'guardian','famsup','schoolsup', 'internet','paid','higher','nursery', 'ses_score'], axis = 1)

In [16]:
df['ses_status'].value_counts()

Medium    632
High      251
Low       161
Name: ses_status, dtype: int64

In [17]:
df.head(5)

school sex  age  reason  traveltime  studytime  failures activities  \
0     GP   F   18  course           2          2         0         no   
1     GP   F   17  course           1          2         0         no   
2     GP   F   15   other           1          2         3         no   
3     GP   F   15    home           1          3         0        yes   
4     GP   F   16    home           1          2         0         no   

  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  Subject  \
0       no       4         3      4     1     1       3         6        1   
1       no       5         3      3     1     1       3         4        1   
2       no       4         3      2     2     3       3        10        1   
3      yes       3         2      2     1     1       5         2        1   
4       no       4         3      2     1     2       5         4        1   

   Final_Score ses_status  
0          6.0     Medium  
1          5.0     Medium  
2          8.0       High  
3         15.0       High  
4          9.0     Medium

## Derived Feature - avg_alc

In [18]:
#For alcohol let's consider categories:
#1 : very low
#2 : low
#3 : moderate
#4 : high
#5 : very high
df['avg_alc'] = (df['Dalc'] + df['Walc']) / 2

In [19]:
df.head(5)

school sex  age  reason  traveltime  studytime  failures activities  \
0     GP   F   18  course           2          2         0         no   
1     GP   F   17  course           1          2         0         no   
2     GP   F   15   other           1          2         3         no   
3     GP   F   15    home           1          3         0        yes   
4     GP   F   16    home           1          2         0         no   

  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  Subject  \
0       no       4         3      4     1     1       3         6        1   
1       no       5         3      3     1     1       3         4        1   
2       no       4         3      2     2     3       3        10        1   
3      yes       3         2      2     1     1       5         2        1   
4       no       4         3      2     1     2       5         4        1   

   Final_Score ses_status  avg_alc  
0          6.0     Medium      1.0  
1          5.0     Medium      1.0  
2          8.0       High      2.5  
3         15.0       High      1.0  
4          9.0     Medium      1.5

In [20]:
#dropping the features used in calculating ses_status
df = df.drop(['Dalc', 'Walc'], axis = 1)

### Derived Feature - Social_Score

In [21]:
label = {'no': 0, 'yes': 1}
df['romantic'] = df['romantic'].map(label)

In [22]:
# Perform the ANOVA test for go out with friends vs. Final Score
goout_anova_result = f_oneway(
    df[df['goout'] == 1]['Final_Score'],
    df[df['goout'] == 2]['Final_Score'],
    df[df['goout'] == 3]['Final_Score'],
    df[df['goout'] == 4]['Final_Score'],
    df[df['goout'] == 5]['Final_Score']
)

# Perform the ANOVA test for family relationship vs. Final Score
fam_anova_result = f_oneway(
    df[df['famrel'] == 1]['Final_Score'],
    df[df['famrel'] == 2]['Final_Score'],
    df[df['famrel'] == 3]['Final_Score'],
    df[df['famrel'] == 4]['Final_Score'],
    df[df['famrel'] == 5]['Final_Score']
)

# Perform the ANOVA test for Romantic relationship vs. Final Score
rom_anova_result = f_oneway(
    df[df['romantic'] == 1]['Final_Score'],
    df[df['romantic'] == 0]['Final_Score'] 
)

In [23]:
print("ANOVA test result:")

# Print the ANOVA test result for go out with friends vs final score
print("go out with friend vs. Final Score:  F-statistic =", goout_anova_result.statistic, " p =", goout_anova_result.pvalue)

# Print the ANOVA test result for family relationship vs final score
print("family relationship vs. Final Score: F-statistic =", fam_anova_result.statistic, "p =", fam_anova_result.pvalue)

# Print the ANOVA test result for relationship vs final score
print("romantic relationship vs. Final Score:F-statistic =", rom_anova_result.statistic, "  p =", rom_anova_result.pvalue)


ANOVA test result:
go out with friend vs. Final Score:  F-statistic = 8.273167699778869  p = 1.4369456295221047e-06
family relationship vs. Final Score: F-statistic = 1.7828736587307605 p = 0.1299868897439774
romantic relationship vs. Final Score:F-statistic = 8.84918127133717   p = 0.002999795908015832


#### Going Out with Friends vs. Final Score:

#### F- statistic = 8.27316769
#### p-value (p) = 1.4369456295221047e-06
#### Interpretation: The small p-value (close to 0) indicates that there is a statistically significant difference in the means of the Final Score across different groups defined by the variable "go out with friend." In other words, "go out with friend" is likely to have a significant impact on the Final Score.

#### Family Relationships vs. Final Score:

#### F- statistic = 1.782873
#### p-value (p) = 0.129986
#### Interpretation: In this case, the p-value is greater than 0.05, indicating that there is no statistically significant difference in the means of the Final Score across different groups defined by the variable "family relationship." Therefore, the evidence does not support a significant impact of "family relationship" on the Final Score.


#### Romantic Relationship vs. Final Score:

#### F- statistic = 8.84918
#### p-value (p) = 0.0029997
#### Interpretation: The p-value is less than 0.05, indicating that there is a statistically significant difference in the means of the Final Score across different groups defined by the variable "romantic relationship." Thus, "romantic relationship" appears to have a significant impact on the Final Score.

In [24]:
def calculate_social_score(row):
    friends_weight = 0.5
    romantic_weight = 0.35
    family_weight = 0.15

    # Calculate the social score using weighted average
    social_score = (row['famrel'] * family_weight +
                    row['goout'] * friends_weight +
                    row['romantic'] * romantic_weight)

    # Round the social_score to the nearest integer and return it
    return int(round(social_score))

In [25]:
# Apply the function to create the 'social_score' feature in the DataFrame
df['social_score'] = df.apply(calculate_social_score, axis=1)

In [26]:
# Define the categories for 'social_score'
df['social_score_category'] = pd.cut(df['social_score'], bins=[0, 2, 3, 5], labels=['Very Bad', 'Neutral', 'Excellent'])

In [27]:
df = df.drop(['famrel', 'romantic', 'goout', 'social_score'], axis = 1)

In [28]:
df

school sex  age  reason  traveltime  studytime  failures activities  \
0        GP   F   18  course           2          2         0         no   
1        GP   F   17  course           1          2         0         no   
2        GP   F   15   other           1          2         3         no   
3        GP   F   15    home           1          3         0        yes   
4        GP   F   16    home           1          2         0         no   
...     ...  ..  ...     ...         ...        ...       ...        ...   
1039     MS   F   19  course           1          3         1        yes   
1040     MS   F   18  course           1          2         0         no   
1041     MS   F   18  course           2          2         0        yes   
1042     MS   M   17  course           2          1         0         no   
1043     MS   M   18  course           3          1         0         no   

      freetime  health  absences  Subject  Final_Score ses_status  avg_alc  \
0            3       3         6        1          6.0     Medium      1.0   
1            3       3         4        1          5.0     Medium      1.0   
2            3       3        10        1          8.0       High      2.5   
3            2       5         2        1         15.0       High      1.0   
4            3       5         4        1          9.0     Medium      1.5   
...        ...     ...       ...      ...          ...        ...      ...   
1039         4       5         4        0         10.0        Low      1.5   
1040         3       1         4        0         15.0       High      1.0   
1041         1       5         6        0         10.0        Low      1.0   
1042         4       2         6        0         10.0     Medium      3.5   
1043         4       5         4        0         11.0     Medium      3.5   

     social_score_category  
0                  Neutral  
1                 Very Bad  
2                 Very Bad  
3                 Very Bad  
4                 Very Bad  
...                    ...  
1039              Very Bad  
1040               Neutral  
1041              Very Bad  
1042               Neutral  
1043              Very Bad  

[1044 rows x 16 columns]

In [29]:
df.to_csv('Preprocessed-Data.csv')